### ISSUE TITLE
~~OpenAI's Gym is causing issues with installation~~ Mystery issue is now causing problems on install

### ISSUE THREAD

**Issue by JD-The-65th** [minerllabs/minerl#707]
### IGNORE THIS, This issue has been fixed. See below.
TL;DR: Update the Unmaintained _gym_ package to a version higher than 0.20, or update Gym to the now maintained _gymnasium_ package

### The Issue

OpenAI's _gym_ package now causes an error when installing on versions lower than v0.20. This error makes MineRL uninstallable. Here's the error in question:

```Collecting gym==0.18.0
  Downloading gym-0.18.0.tar.gz (1.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 2.4 MB/s eta 0:00:00m eta 0:00:010:01:010m
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [1 lines of output]
      error in gym setup command: 'extras_require' must be a dictionary whose values are strings or lists of strings containing valid project/version requirement specifiers.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
```

My python version is 3.10.9
In my testing, this error did not occur on versions 0.20 and later. I've identified 2 solutions.

### Solution 1

Update the (now unmaintained) gym package to 0.20 or later, or

### Solution 2 (objectively better imho)

Update to the now maintained _gymnasium_ package from the Farama Foundation. It works as a Gym dropin,

`import gymnasium as gym`

but ypu will have to rewrite some of the Gym code. More info can be found at  https://www.gymlibrary.dev/

**Comment by JD-The-65th** [comment]
This _might_ be solvable by limiting the python version to another version, but I'm not sure. That requires more testing, and it is 1:51 AM as I'm writing this so I don't have the time, nor the patience.

**Comment by Miffyli** [comment]
Heya. Yup this is sad and a known issue... The issue is that MineRL mutilated the original Gym code in ways that were not "meant to be", so already upgrading to 0.20 is a breaking change for MineRL :/. It has been time since I looked into properly fixing it, but it would mean refactoring the environments quite a bit.

~~Can you detail your Python versions / setup so I can replicate this? I have been meaning to fix this but I have not been able to create a Python env where this fails 😅 (might have installed packages that changed setuptools to fix it). I think I will be forking gym 0.19, fixing it and making it the default for MineRL.~~

Edit: I am blind, you included the python version. Let me try with this to see if I can fix things.

**Comment by JD-The-65th** [comment]
Forking Gym and applying the changes to 0.19 is a good solution I didn't consider. I don't have the best of experience,  but I'd say limiting the python version might stall the issue but cause other ones later down the road, so I say forking gym and applying the changes there would be better.

**Comment by Miffyli** [comment]
@JD-The-65th  Applied the fix now. Could you give it a try to see if it works? 

It is really nasty now that I think of it, but I could not figure out better one... Debugging this was fun as well, because I could not replicate it with any python version on my WSL, regardless of conda environment I used.

**Comment by JD-The-65th** [comment]
Good news is, I'm not getting a Gym related issue! Bad news is, MCP-Reborn is causing an issue now. Same py version, and I'm forcing jdk8 to be used. Error is too long to be copied into this message, so I made it into a txt file
[error.txt](https://github.com/minerllabs/minerl/files/11595664/error.txt)

**Comment by JD-The-65th** [comment]
I'm gonna test to see if this works in colab, and if it does, I'm gonna close this issue and open another one. If it doesn't, im keeping this one open.

**Comment by JD-The-65th** [comment]
I got a totally different error from the one I got locally, this time in colab (free) using the vpt colab notebook. This is all it said, so I don't have any leads.

```
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Collecting git+https://github.com/minerllabs/minerl@v1.0.0
  Cloning https://github.com/minerllabs/minerl (to revision v1.0.0) to /tmp/pip-req-build-ps7hphgd
  Running command git clone --filter=blob:none --quiet https://github.com/minerllabs/minerl /tmp/pip-req-build-ps7hphgd
  Running command git checkout -b v1.0.0 --track origin/v1.0.0
  Switched to a new branch 'v1.0.0'
  Branch 'v1.0.0' set up to track remote branch 'v1.0.0' from 'origin'.
  Resolved https://github.com/minerllabs/minerl to commit c0e07f204ae45982536dba85701c094695641e0b
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

---------------------------------------------------------------------------

ModuleNotFoundError                       Traceback (most recent call last)

[<ipython-input-3-4b1a49133b7f>](https://localhost:8080/#) in <cell line: 2>()
      1 get_ipython().system('pip3 install --upgrade git+https://github.com/minerllabs/minerl@v1.0.0')
----> 2 import minerl

ModuleNotFoundError: No module named 'minerl'
```

**Comment by Miffyli** [comment]
Installing `git+https://github.com/minerllabs/minerl@v1.0.0` will probably fail. Try `git+https://github.com/minerllabs/minerl`

Edit: I also applied a second fix (that's what I get for not trying the full installation pipeline... 😅 ). Try now

**Comment by JD-The-65th** [comment]
> Installing `git+https://github.com/minerllabs/minerl@v1.0.0` will probably fail. Try `git+https://github.com/minerllabs/minerl`
> 
> Edit: I also applied a second fix (that's what I get for not trying the full installation pipeline... sweat_smile ). Try now

Locally I get the same issue from error.txt I linked above. Testing in colab now.

**Comment by JD-The-65th** [comment]
It built, (taking an absurdly long 20 minutes to in colab tho) so I'm closing this as completed.
